src/helpers.c

```diff
@@ -149,130 +149,116 @@ void concat_str_array(char **output, char *const argv[])
 	}
 }
 
-int system_argv(char *const argv[])
+int system_argv_pipe(char *const lhscmd[], char *const rhscmd[])
 {
-	int child_exit_status;
-	pid_t pid;
-	int status = -1;
+	return system_argv_pipe_fd(-1, -1, lhscmd, -1, -1, rhscmd);
+}
 
-	pid = fork();
+int system_argv_pipe_fd(int lnewstdinfd, int lnewstderrfd, char *const lhscmd[],
+			int rnewstdoutfd, int rnewstderrfd, char *const rhscmd[])
+{
+	pid_t monitorpid = fork();
+	if (monitorpid == -1) {
+		LOG(NULL, "Failed to create child process to monitor pipe between", "command %s and command %s", lhscmd[0], rhscmd[0]);
+		return -1;
+	} else if (monitorpid == 0) {
+		pipe_monitor(lnewstdinfd, lnewstderrfd, lhscmd, rnewstdoutfd, rnewstderrfd, rhscmd);
+	}
+	return wait_process_terminate(monitorpid);
+}
 
-	if (pid == 0) { /* child */
-		execvp(*argv, argv);
-		LOG(NULL, "This line must not be reached", "");
-		assert(0);
-	} else if (pid < 0) {
-		LOG(NULL, "Failed to fork a child process", "");
+void pipe_monitor(int lnewstdinfd, int lnewstderrfd, char *const lhscmd[],
+		  int rnewstdoutfd, int rnewstderrfd, char *const rhscmd[])
+{
+	int pipefd[2];
+	if (pipe(pipefd) == -1) {
+		LOG(NULL, "Failed to create a pipe between", "command %s and command %s", lhscmd[0], rhscmd[0]);
 		assert(0);
-	} else {
-		pid_t ws = waitpid(pid, &child_exit_status, 0);
-
-		if (ws == -1) {
-			LOG(NULL, "Failed to wait for child process", "");
-			assert(0);
-		}
-
-		if (WIFEXITED(child_exit_status)) {
-			status = WEXITSTATUS(child_exit_status);
-		} else {
-			LOG(NULL, "Child process didn't exit", "");
-			assert(0);
-		}
+	}
 
-		if (status != 0) {
-			char *cmdline = NULL;
+	pid_t lhspid = system_argv_fd_nowait(lnewstdinfd, pipefd[1], lnewstderrfd, &pipefd[0], lhscmd);
```

Comment:
This works, but I would have coded system_argv_fd_nowait to have its 4th parameter as an int, rather than a pointer to int, and used -1 to indicate there is nothing to close. Passing pointers around makes it harder to reason about code.

Reply:
Good point.  I'll fix this :)